FuelLabs/fuel-indexer
# enhancement: implement instantiation methods for entity types

Closes #951 and #963.

### Description
This PR adds a `::new()` method to entity types that automatically calculates an ID for the user so that they do not have to come up with their own way to create one. It also adds a `::get_or_create()` method that will load a record from the DB if present or create the struct as described above.

### Testing steps

The `hello-world` example has been changed to illustrate the new instantiation methods.
1. Ensure the test DB is cleared and that the `hello-world` WASM module has been freshly built.
2. `cargo run -p hello-world-node`
3. `cargo run --bin fuel-indexer -- run --run-migrations --manifest examples/hello-world/hello-indexer/hello_indexer.manifest.yaml --verbose`
4. `cargo run -p hello-world-data`
5. Check the output of the indexer; you should see that the original and `::new()` methods insert data into the DB, and that `::get_or_create()` loads a record from the database.

### Changelog

- Add a number of `HashSet`s for checking whether a field type implements the requisite traits
- Add a helper function for creating the necessary method implementations
